### PR TITLE
Update from Rails 6 to Rails 7

### PR DESCRIPTION
**Pull request by KludgeKML**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

https://trello.com/c/drr3pXQb/1147-rails-7-update-authenticating-proxy

### CODE REVIEW

File: config/application.rb [hunk 1]
```diff
@@ -1,22 +1,34 @@
...
+require "active_job/railtie"
+# require "active_record/railtie"
+# require "active_storage/engine"
 require "action_controller/railtie"
+# require "action_mailer/railtie"
+# require "action_mailbox/engine"
+# require "action_text/engine"
+require "action_view/railtie"
+# require "action_cable/engine"
+require "rails/test_unit/railtie"
```
Comment (1pretz1) [inline config/application.rb:15]: Do we need these extra frameworks? I note that we don't use ActiveJob, Views or test_unit

File: config/environments/development.rb [hunk 2]
```diff
@@ -45,14 +48,6 @@
...
-  # Use an evented file watcher to asynchronously detect changes in source code,
-  # routes, locales, etc. This feature depends on the listen gem.
-  config.file_watcher = ActiveSupport::EventedFileUpdateChecker
-
   # Uncomment if you wish to allow Action Cable access from any origin.
   # config.action_cable.disable_request_forgery_protection = true
-
-  config.hosts += [
-    "authenticating-proxy.dev.gov.uk",
-  ]
```
Comment (1pretz1) [inline config/environments/development.rb:57]: I believe we'll want to keep the hosts config as this app does run in Docker and connecting via the host is mentioned in the README: https://github.com/alphagov/authenticating-proxy/blob/main/README.md

File: config/environments/test.rb [hunk 3]
```diff
@@ -8,12 +8,13 @@
...
   config.cache_classes = true
 
-  # Do not eager load code on boot. This avoids loading your whole application
-  # just for the purpose of running a single test. If you are using a tool that
-  # preloads Rails for running tests, you may have to set it to true.
-  config.eager_load = false
+  # Eager loading loads your whole application. When running a single test locally,
+  # this probably isn't necessary. It's a good idea to do in a continuous integration
+  # system, or in some way before deploying your code.
+  config.eager_load = ENV["CI"].present?
```
Comment (1pretz1) [inline config/environments/test.rb:17]: I wonder if there is a speed benefit if eager_load is set to true in our CI environments (doesn't really matter for this app as it's tiny there's like 3 tests)

File: Gemfile.lock [hunk 4]
```diff
@@ -68,42 +68,23 @@ GEM
       tzinfo (~> 2.0)
```
Comment (1pretz1) [inline Gemfile.lock:68]: Are these bumps needed due to dependency issues or just for good measure?

**Review by 1pretz1 — COMMENTED** [review]
Changes look good - just left a few comments. As I couldn't comment on the commit, we probably don't need the empty commit: https://github.com/alphagov/authenticating-proxy/pull/313/commits/2a056dd32953c89802f005428886f6ee8985c149, although thanks for highlighting there were no deprecation notices 👍 (maybe this information could just live in the PR description?) 

We'd probably want to deploy this to integration before merging just to check everything's working as expected and we can access the draft stack.

File: Gemfile.lock (hunk 4 above)
Comment (KludgeKML) [reply Gemfile.lock:68]: Part of the process we're following from publishing includes "bump other gems" (ie I think we're trying to get ahead of dependabot a bit as we bump rails).

File: config/environments/test.rb (hunk 3 above)
Comment (KludgeKML) [reply config/environments/test.rb:17]: Don't know - this is the default change put in by the rails update, so per the process I think we're supposed to keep it unless it's going to cause trouble. I assume so we keep as close to base Rails 7 config as possible.

File: config/environments/development.rb (hunk 2 above)
Comment (KludgeKML) [reply config/environments/development.rb:57]: Yes, good point. I'll revert.

**Comment by KludgeKML** [comment]
I initially put in the empty commit because I thought it would be useful for history if all apps had the same set of commits, but I think no-one else is doing this, so even that flimsy justification isn't there. I'll squash it and note in the PR.

File: Gemfile.lock (hunk 4 above)
Comment (KludgeKML) [reply Gemfile.lock:68]: Oh, in fact in this commit they're dependencies of the rails update.

File: config/application.rb (hunk 1 above)
Comment (KludgeKML) [reply config/application.rb]: Fixed.

File: config/environments/development.rb (hunk 2 above)
Comment (KludgeKML) [reply config/environments/development.rb]: Fixed.

File: Gemfile.lock (hunk 4 above)
Comment (1pretz1) [reply Gemfile.lock:68]: That makes sense (getting ahead of Dependabot) 👍

**Review by 1pretz1 — APPROVED** [review]

**Pull request merged**